### PR TITLE
Alerting: grafana managed group names

### PR DESCRIPTION
**Pull request by gillesdemey**
**What this PR does / why we need it**:

This PR introduces proper support for arbitrary group names in Grafana managed alerts. Previously all Grafana managed rules were placed in to their own group of exactly one rule.

In an effort to unify the behavior between Prometheus rules and Grafana managed rules and to comply with the [Prometheus Alert Generator Compliance Specification](https://github.com/prometheus/compliance/blob/main/alert_generator/specification.md) this PR adds the following:

### An additional tab to view Grafana managed rules in their group hierarchy

The existing grouped view has been renamed to "List" since it flattens the view hierarchy for Grafana managed rules.
A new tab has been added called "grouped"

image

image

### Rules can be moved to other groups

In addition to moving rules between namespaces you can now also move rules to another group within the same namespace. The existing `saveGrafanaRule` has been substantially refactored.

### Additional information is included

To better communicate the new alerting paradigm I've included some inline documentation on folder and group. I'll need some input from @achatterjee-grafana so we can refine the content.

image

Since the rule interval will also determine the interval of all other rules in the group we have to explicitly call this out.

image

I'm not quite sure if the interval configuration is in the right place – this is something I'd like @jessover9000's opinion on.

**Which issue(s) this PR fixes**:

Related to #45881

**Special notes for your reviewer**:

I still have to do a quick review myself as I've touch on quite a few files and probably have some TODOs sprinkled around and some variable and function names could be improved – feel free to help bikeshed :)

### CODE REVIEW

File: public/app/features/alerting/unified/utils/rulerClient.ts [hunk 1]
```diff
@@ -157,65 +141,106 @@ export function getRulerClient(rulerConfig: RulerDataSourceConfig): RulerClient
...
+    } else {
+      // we're moving a rule to either a different group or namespace
+      return moveGrafanaRule(namespace, groupSpec, existingRule, newRule);
+    }
+  };
 
-    // set group name to rule name, but be super paranoid and check that this group does not already exist
-    const groupName = getUniqueGroupName(values.name, targetFolderGroups);
-    formRule.grafana_alert.title = groupName;
+  // add a rule to a namespace and group
```
Comment (peterholmberg) [inline public/app/features/alerting/unified/utils/rulerClient.ts:171]: maybe this comment is ambiguous 🙂

File: public/app/features/alerting/unified/utils/rulerClient.ts [hunk 2]
```diff
@@ -157,65 +141,106 @@ export function getRulerClient(rulerConfig: RulerDataSourceConfig): RulerClient
...
+    await setRulerRuleGroup(rulerConfig, namespace, payload);
 
-      return { uid: newUid, ruleSourceName: 'grafana' };
-    } else {
-      throw new Error('Failed to fetch created rule.');
+    // TODO figure out how to get the newly created UID :(
+    return { uid: '', ruleSourceName: GRAFANA_RULES_SOURCE_NAME };
+  };
+
+  // we can't move the rule atomically so we have to
```
Comment (peterholmberg) [inline public/app/features/alerting/unified/utils/rulerClient.ts:194]: nit: spelling

**Review by peterholmberg — COMMENTED** [review]
Just some minor nits. But I have two questions.

1. I see "default" mentioned in some places. Is this referring to a default folder that we talked about?
2. How are current Grafana managed rules migrated to fit within a group?

File: public/app/features/alerting/unified/utils/rulerClient.ts (hunk 2 above)
Comment (gillesdemey) [reply public/app/features/alerting/unified/utils/rulerClient.ts:194]: I meant so clarify here that there isn't a single atomic operation we can perform (atomically), but I can change the wording :D

File: public/app/features/alerting/unified/utils/rulerClient.ts (hunk 1 above)
Comment (gillesdemey) [reply public/app/features/alerting/unified/utils/rulerClient.ts:171]: Definitely, will remove

File: public/app/features/alerting/unified/components/rules/RulesTable.tsx [hunk 3]
```diff
@@ -137,8 +136,18 @@ function useColumns(showSummaryColumn: boolean, showGroupColumn: boolean) {
         // eslint-disable-next-line react/display-name
         renderCell: ({ data: rule }) => {
           const { namespace, group } = rule;
-          const { rulesSource } = namespace;
-          return isCloudRulesSource(rulesSource) ? `${namespace.name} > ${group.name}` : namespace.name;
+
+          // ungrouped rules are rules that are in the "default" group name
+          const isUngrouped = group.name === 'default';
```
Comment (konrad147) [inline public/app/features/alerting/unified/components/rules/RulesTable.tsx:141]: This code looks almost the same as https://github.com/grafana/grafana/blob/81a5df591caa1fe0097b34e7b142297b3e50f750/public/app/features/alerting/unified/components/rules/RulesGroup.tsx#L122:L129
Maybe we can extract a common component for them?

File: public/app/features/alerting/unified/utils/rulerClient.ts [hunk 4]
```diff
@@ -157,65 +141,105 @@ export function getRulerClient(rulerConfig: RulerDataSourceConfig): RulerClient
...
+
+    // make sure our updated alert has the same UID as before
+    const uid = existingRule.rule.grafana_alert.uid;
+    newRule.grafana_alert.uid = uid;
+
+    // create the new array of rules we want to send to the group
+    const newRules = existingRule.group.rules
+      .filter((rule): rule is RulerGrafanaRuleDTO => isGrafanaRulerRule(rule))
+      .filter((rule) => rule.grafana_alert.uid !== uid)
+      .concat(newRule as RulerGrafanaRuleDTO); // TODO why type cast this?
```
Comment (konrad147) [inline public/app/features/alerting/unified/utils/rulerClient.ts:234]: TODO

File: public/app/features/alerting/unified/utils/rulerClient.ts [hunk 5]
```diff
@@ -90,13 +79,8 @@ export function getRulerClient(rulerConfig: RulerDataSourceConfig): RulerClient
   };
 
   const deleteRule = async (ruleWithLocation: RuleWithLocation): Promise<void> => {
-    const { ruleSourceName, namespace, group, rule } = ruleWithLocation;
-    // in case of GRAFANA, each group implicitly only has one rule. delete the group.
-    if (isGrafanaRulesSource(ruleSourceName)) {
-      await deleteRulerRulesGroup(rulerConfig, namespace, group.name);
-      return;
-    }
-    // in case of CLOUD
```
Comment (konrad147) [inline public/app/features/alerting/unified/utils/rulerClient.ts:99]: Nice refactoring in this file 🚀

File: public/app/features/alerting/unified/components/rules/RulesFilter.tsx [hunk 6]
```diff
@@ -13,10 +13,15 @@ import { Stack } from '@grafana/experimental';
 
 const ViewOptions: SelectableValue[] = [
   {
-    icon: 'folder',
-    label: 'Groups',
+    icon: 'list-ul',
+    label: 'List',
     value: 'group',
```
Comment (konrad147) [inline public/app/features/alerting/unified/components/rules/RulesFilter.tsx:18]: Can we change the `group` to `list`? Or we should keep it due to compatibility reasons?

**Comment by konrad147** [comment]
I wonder if we also should unify the way we allow to edit rule groups? 
For cloud rules, the edit button opens the rule group's edit modal
image
Whereas, for Grafana Managed rules the same button opens the folder edit view. Moreover, this rule group edit modal gives us a better place to set the evaluation interval.
I think we don't need to do this in this PR. It's just something we may want to consider doing in the future

File: public/app/features/alerting/unified/components/rules/RulesTable.tsx (hunk 3 above)
Comment (gillesdemey) [reply public/app/features/alerting/unified/components/rules/RulesTable.tsx:141]: Good idea!

File: public/app/features/alerting/unified/utils/rulerClient.ts (hunk 4 above)
Comment (gillesdemey) [reply public/app/features/alerting/unified/utils/rulerClient.ts:234]: It was being typecast before, I didn't dig too deep in to the reason why – might dig a little deeper if I have the time.

File: public/app/features/alerting/unified/utils/rulerClient.ts [hunk 7]
```diff
@@ -157,65 +141,105 @@ export function getRulerClient(rulerConfig: RulerDataSourceConfig): RulerClient
...
-        if (freshExisting && freshExisting.namespace !== folder.title) {
-          await deleteRule(freshExisting);
-        }
-      }
+    await setRulerRuleGroup(rulerConfig, namespace, payload);
 
-      return { uid: newUid, ruleSourceName: 'grafana' };
-    } else {
-      throw new Error('Failed to fetch created rule.');
+    // TODO figure out how to get the newly created UID :(
```
Comment (gillesdemey) [inline public/app/features/alerting/unified/utils/rulerClient.ts:189]: **question:** sadly we don't have the means to find the UID, we just grabbed the UID of the only rule in the created group previously but we can no longer do that here. And suggestions or ideas more than welcome!

**Comment by gillesdemey** [comment]
> I wonder if we also should unify the way we allow to edit rule groups?

Yes absolutely, I was initially thinking of including it in the PR but it's grown quite large and I think we should tackle it incrementally.

**Pull request closed without merging**